cachi2/core/rooted_path.py

```diff
@@ -70,9 +70,8 @@ def subpath_from_root(self) -> Path:
         return self._path.relative_to(self._root)
 
     def __eq__(self, other: object) -> bool:
-        # Note: __eq__ is implemented mainly for unit tests that do assert_called_with() or similar
         if not isinstance(other, RootedPath):
-            return NotImplemented
+            raise NotImplementedError(f"__eq__ not implemented for {type(other).__name__}")
```

Comment:
Well, I thought, that raising an exception wouldn't be better than just returning **False** while comparing _RootedPath_ with _pathlib.Path_ for example. I might save some debugging time.

Reply:
>  I thought, that raising an exception wouldn't be better than just returning False while comparing RootedPath with pathlib.Path for example. I might save some debugging time.

So I read the docs on `NotImplementedError` exception and I have to retract on my previous answer. The intended use case is for base classes and abstract methods, which makes sense during development. However, while you're perfectly fine to use it anyway you wish, there are a couple of things to consider which I believe suggest against using it in this case:

1. Deliberately raising an exception on a production code instead of wrapping it by a nice user-friendly error is an antipattern which means `NotImplemented` is actually a better choice here
2. Pathlib itself uses `NotImplemented` instead of an exception: https://github.com/python/cpython/blob/d308d33e098d8e176f1e5169225d3cf800ed6aa1/Lib/pathlib/__init__.py#L203

Reply:
That's a good argument. I just kind of don't understand what the extra value of `NotImplemented`.
```
In [1]: from cachi2.core.rooted_path import RootedPath

In [2]: x = RootedPath("/home/cachi2")

In [3]: from pathlib import Path

In [4]: y = Path("/home/cachi2")

In [5]: x == y
Out[5]: False
```

Reply:
Well, you will need to go through [1] and [2] yourself if you want deep understanding ;), but the gist here is that `NotImplemented` is a special Python builtin type which is used solely with rich comparisons (i.e. usage of methods `__eq__`, `__lt__`, `__gt__`, etc. as a replacement of the old mechanism implemented by `__cmp__`) and nothing else, in other words, you wouldn't use `NotImplemented` for a random function/method that you declared, but deferred the implementation  (for that you'd use an exception). IIUC `NotImplemented` serves the purpose of chain hierarchy of fallback behaviours ultimately leading to a `TypeError` which I understand you'd be kind of after from your example.

What I think is important to understand here is that while ultimately you'd still get a `TypeError`, Python is trying hard to defer that logic to the worst case scenario and tries other possibilities first, ridding you of the need to put error recovering logic absolutely everywhere.

[1] https://docs.python.org/3/library/constants.html#NotImplemented
[2] https://docs.python.org/3/reference/datamodel.html#object.__eq__